src/GafferScene/MeshTangents.cpp

```diff
@@ -151,12 +187,38 @@ IECore::ConstObjectPtr MeshTangents::computeProcessedObject( const ScenePath &pa
 
 	std::string uvSet = uvSetPlug()->getValue();
 	std::string position = positionPlug()->getValue();
+	std::string normal = normalPlug()->getValue();
 	bool ortho = orthogonalPlug()->getValue();
+	bool leftHanded = leftHandedPlug()->getValue();
+	MeshAlgo::MeshTangentsMode mode = (MeshAlgo::MeshTangentsMode) modePlug()->getValue();
 
 	std::string uTangent = uTangentPlug()->getValue();
 	std::string vTangent = vTangentPlug()->getValue();
 
-	std::pair<PrimitiveVariable, PrimitiveVariable> tangentPrimvars = MeshAlgo::calculateTangents( mesh, uvSet, ortho, position);
+	std::pair<PrimitiveVariable, PrimitiveVariable> tangentPrimvars;
+
+	if ( mode == MeshAlgo::MeshTangentsMode::UV )
+	{
+		tangentPrimvars = MeshAlgo::calculateTangentsFromUV( mesh, uvSet, position, ortho, leftHanded );
+	}
+	else if ( mode == MeshAlgo::MeshTangentsMode::FirstEdge )
+	{
+		tangentPrimvars = MeshAlgo::calculateTangentsFromFirstEdge( mesh, position, normal, ortho, leftHanded  );
+	}
+	else if ( mode == MeshAlgo::MeshTangentsMode::TwoEdges )
+	{
+		tangentPrimvars = MeshAlgo::calculateTangentsFromTwoEdges( mesh, position, normal, ortho, leftHanded  );
+	}
+	else if ( mode == MeshAlgo::MeshTangentsMode::PrimitiveCentroid )
+	{
+		tangentPrimvars = MeshAlgo::calculateTangentsFromPrimitiveCentroid( mesh, position, normal, ortho, leftHanded  );
+	}
+	else
+	{
+		std::string e = boost::str( boost::format( "MeshTangents : Invalid MeshTangentMode \"%s\"." ) % mode );
```

Comment:
If I recall correctly, you might be able to get away with it in boost::format, but %s isn't a logical specifier for an enum/int. If we enforce the appropriate min/max on the plug constructor, we won't need this block anyway...